### PR TITLE
Refactor executors again

### PR DESCRIPTION
**Pull request by rabernat**
This was a little coding project while stuck in a plane for a few hours. Here I am basically just copying over pangeo-forge-recipes' latest executors, which we spent a lot of time optimizing.

Fixes
- #110 
- #92 (I hope; have not actually tested directly yet)

Locally I have pywren tests failing with a message `TypeError: cannot pickle 'weakref' object`. Was not able to debug this quickly as I'm not too familiar with pywren.

### CODE REVIEW

**Comment by tomwhite** [comment]
> Locally I have pywren tests failing

There has not been a pywren release for over 3 years, since it has now become lithops. I wonder if we should remove or disable the pywren executor for this reason.

**Comment by rabernat** [comment]
Ok so in fb8b1ee I just removed the pywren executor. If it is unmaintained, it's not feasible for us to find workarounds to keep supporting it. @tomwhite - are you using the pywren executor these days?

I had a look at lithops, and it looks like it should be pretty straightforward to support. It would be nice to have a serverless option. If someone wants to implement and maintain a lithops executor, I would be happy to support that effort.

File: tests/test_rechunk.py [hunk 1]
```diff
@@ -361,10 +354,14 @@ def test_rechunk_array(
     assert target_array.chunks == tuple(target_chunks_list)
     assert dict(source_array.attrs) == dict(target_array.attrs)
 
+    # print(rechunked.plan[0].dask)
     result = rechunked.execute()
     assert isinstance(result, zarr.Array)
-    a_tar = dsa.from_zarr(target_array)
-    assert dsa.equal(a_tar, 1).all().compute()
+    a_tar = np.asarray(result)
+    print(a_tar.shape, a_tar.dtype)
```
Comment (TomAugspurger) [inline tests/test_rechunk.py:361]: (nitpick): left a debug print here. And can / should the `assert` below be uncommented?

File: setup.py [hunk 2]
```diff
@@ -43,14 +43,14 @@
         "Topic :: Database",
         "Topic :: Scientific/Engineering",
         "License :: OSI Approved :: MIT License",
-        "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
```
Comment (TomAugspurger) [inline setup.py:48]:
```suggestion
        "Programming Language :: Python :: 3.9",
        "Programming Language :: Python :: 3.10",
```

File: rechunker/types.py [hunk 3]
```diff
@@ -8,8 +9,11 @@
     Optional,
     Tuple,
     TypeVar,
+    Union,
 )
 
+from mypy_extensions import NamedArg
```
Comment (TomAugspurger) [inline rechunker/types.py:15]: Is this a necessary runtime dependency? Might need to add it to `install_requires` in the `setup.py` / `setup.cfg`.

**Comment by rabernat** [comment]
For some reason the test workflow is not running.

**Comment by TomAugspurger** [comment]
GitHub actions was having issues earlier, but nothing is reported at https://www.githubstatus.com/ at the moment.

**Comment by TomAugspurger** [comment]
Oh, https://github.com/pangeo-data/rechunker/actions/runs/2029808478 shows

```
The workflow is not valid. .github/workflows/ci.yaml (Line: 9, Col: 7): Unexpected value 'fast-fail'
```

**Pull request merged**